File: kernel/arch/x64/smp.rs
use super::ioapic;
use crate::arch::utils::PhyPtr;
use core::mem::size_of;
use core::option::Option;

const MP_BASETABLE_PROCESSOR_ENTRY: u8 = 0;
const MP_BASETABLE_BUS_ENTRY: u8 = 1;
const MP_BASETABLE_IOAPIC_ENTRY: u8 = 2;
const MP_BASETABLE_IOINT_ASSIGN_ENTRY: u8 = 3;
const MP_BASETABLE_LOCALINT_ASSIGN_ENTRY: u8 = 4;

#[repr(C, packed)]
struct FloatPtr {
    signature: [u8; 4], // "_MP_"
    mptable_header_addr: u32,
    length: u8,
    spec_rev: u8,
    checksum: u8,
    info1: u8,
    info2: u8,
    info3: [u8; 3],
}

#[repr(C, packed)]
struct MpTableHeader {
    signature: [u8; 4], // "PCMP"
    base_table_length: u16,
    spec_rev: u8,
    checksum: u8,
    oem_id: [u8; 8],
    product_id: [u8; 12],
    oem_table_pointer: u32,
    oem_table_size: u16,
    entry_count: u16,
    memmaped_localapic_addr: u32,
    extended_table_length: u16,
    extended_table_checksum: u8,
    reserved: u8,
}

#[repr(C, packed)]
struct MpProcessorEntry {
    id: u8,
    localapic_id: u8,
    localapic_ver: u8,
    cpu_flags: u8,
    cpu_signature: u32,
    feature_flags: u32,
    reserved1: u32,
    reserved2: u32,
}

#[repr(C, packed)]
struct MpBusEntry {
    id: u8,
    bus_id: u8,
    type_str: [u8; 6],
}

#[repr(C, packed)]
struct IoApicEntry {
    id: u8,
    ioapic_id: u8,
    ioapic_ver: u8,
    ioapic_flags: u8,
    memmaped_ioapic_addr: u32,
}

#[repr(C, packed)]
struct IoIntAssignEntry {
    id: u8,
    int_type: u8,
    int_flags: u16,
    src_bus_id: u8,
    src_bus_irq: u8,
    dest_ioapic_id: u8,
    dest_ioapic_intin: u8,
}

#[repr(C, packed)]
struct LocalIntAssignEntry {
    id: u8,
    int_type: u8,
    int_flags: u16,
    src_bus_id: u8,
    src_bus_irq: u8,
    dest_localapic_id: u8,
    dest_localapic_intin: u8,
}

/// Looks for MP Floating Pointer Table.
unsafe fn look_for_mp_float_ptr_table() -> Option<*const FloatPtr> {
    let mut p: PhyPtr<u8> = PhyPtr::new(0xf0000);
    for _ in 0..0x10000 {
        let float_ptr: *const FloatPtr = p.as_ptr();
        if (*float_ptr).signature[0] == '_' as u8
            && (*float_ptr).signature[1] == 'M' as u8
            && (*float_ptr).signature[2] == 'P' as u8
            && (*float_ptr).signature[3] == '_' as u8
        {
            return Some(p.as_ptr());
        }

        p = p.add(1);
    }

    None
}

pub unsafe fn init() {
    let float_ptr = look_for_mp_float_ptr_table().expect("failed to detect MP float ptr table");

    let table_header_paddr: PhyPtr<u8> = PhyPtr::new((*float_ptr).mptable_header_addr as usize);
    let table_header: *const MpTableHeader = table_header_paddr.as_ptr();
    if (*table_header).signature[0] != 'P' as u8
        || (*table_header).signature[1] != 'C' as u8
        || (*table_header).signature[2] != 'M' as u8
        || (*table_header).signature[3] != 'P' as u8
    {
        panic!("invalid MP table header");
    }

    let mut entry_ptr = table_header_paddr.vaddr_as_usize() + size_of::<MpTableHeader>();
    for _i in 0..(*table_header).entry_count {
        let entry_type = (entry_ptr as *const u8).read();
        let entry_size = match entry_type {
            MP_BASETABLE_IOAPIC_ENTRY => {
                let ioapic_entry = entry_ptr as *const IoApicEntry;
                if (*ioapic_entry).ioapic_flags != 0 {
                    ioapic::init((*ioapic_entry).memmaped_ioapic_addr as u64);
                }
                return;
            }
            MP_BASETABLE_BUS_ENTRY => size_of::<MpBusEntry>(),
            MP_BASETABLE_PROCESSOR_ENTRY => size_of::<MpProcessorEntry>(),
            MP_BASETABLE_IOINT_ASSIGN_ENTRY => size_of::<IoIntAssignEntry>(),
            MP_BASETABLE_LOCALINT_ASSIGN_ENTRY => size_of::<LocalIntAssignEntry>(),
            _ => panic!("unknown MP table entry type={}", entry_type),
        };

        entry_ptr += entry_size;
    }

    panic!("I/O APIC not found");
}

File: kernel/arch/x64/smp.c
#include <arch.h>
#include <printk.h>
#include <x64/apic.h>
#include <x64/smp.h>
#include <x64/x64.h>

static struct mp_float_ptr *look_for_floatptr_table(
    paddr_t start, paddr_t end) {
    vaddr_t end_vaddr = (vaddr_t) from_paddr(end);
    for (uint32_t *p = from_paddr(start); (vaddr_t) p < end_vaddr; p++) {
        if (*p == MP_FLOATPTR_SIGNATURE) {
            return (struct mp_float_ptr *) p;
        }
    }

    return NULL;
}

void x64_smp_init(void) {
    struct mp_table_header *mptblhdr;
    struct mp_ioapic_entry *ioapic_entry;
    void *entry_ptr;

    struct mp_float_ptr *mpfltptr = look_for_floatptr_table(0xf0000, 0x100000);
    if (mpfltptr == NULL) {
        PANIC("MP table not found");
        return;
    }

    mptblhdr = (struct mp_table_header *) from_paddr(
        (paddr_t) mpfltptr->mptable_header_addr);
    if (mptblhdr->signature != MP_MPTABLE_SIGNATURE) {
        PANIC("invalid MP table");
        return;
    }

    entry_ptr = (void *) ((paddr_t) mptblhdr + sizeof(struct mp_table_header));
    for (int i = 0; i < mptblhdr->entry_count; i++) {
        size_t size;
        uint8_t type = *((uint8_t *) entry_ptr);
        switch (type) {
        case MP_BASETABLE_IOAPIC_ENTRY:
            ioapic_entry = (struct mp_ioapic_entry *) entry_ptr;
            size = sizeof(struct mp_ioapic_entry);
            if (ioapic_entry->ioapic_flags != 0) {
                x64_ioapic_init(ioapic_entry->memmaped_ioapic_addr);
            }
            break;
        case MP_BASETABLE_BUS_ENTRY:
            size = sizeof(struct mp_bus_entry);
            break;
        case MP_BASETABLE_PROCESSOR_ENTRY:
            size = sizeof(struct mp_processor_entry);
            break;
        case MP_BASETABLE_IOINT_ASSIGN_ENTRY:
            size = sizeof(struct mp_ioint_assign_entry);
            break;
        case MP_BASETABLE_LOCALINT_ASSIGN_ENTRY:
            size = sizeof(struct mp_localint_assign_entry);
            break;
        default:
            PANIC("unknown mp table entry: %d", type);
        }

        entry_ptr = (void *) ((paddr_t) entry_ptr + (paddr_t) size);
    }
}
